File: python/treelas/lca.py
"""
Determine the lowest common ancestor
"""


def test_defaultdict():
    from collections import defaultdict

    assert bool() == False
    colors = defaultdict(bool)          # dictionary with default False
    assert colors[0] == False
    assert colors[1] == False


def lca_networkx(G, root, pairs):
    """
    [`networkx.algorithms.lowest_common_ancestor`][nx]
    Implemented according to CLRS page 584 (3rd edition).
    Compare to [epp]

    [nx]: https://github.com/networkx/networkx/master/
    [epp]: https://www.ics.uci.edu/~eppstein/PADS/LCA.py

    """
    from collections import defaultdict
    from networkx.utils import UnionFind, arbitrary_element
    from networkx import dfs_postorder_nodes


    pair_dict = defaultdict(set)
    for u, v in pairs:
        pair_dict[u].add(v)
        pair_dict[v].add(u)

    # Iterative implementation of Tarjan's offline lca algorithm
    # as described in CLRS on page 521.
    uf = UnionFind()
    ancestors = {}
    for node in G:
        ancestors[node] = uf[node]

    colors = defaultdict(bool)  
    for node in dfs_postorder_nodes(G, root):
        colors[node] = True
        for v in pair_dict[node]:
            if colors[v]:
                if (v, node) in pairs:
                    yield (v, node), ancestors[uf[v]]
        if node != root:
            parent = arbitrary_element(G.pred[node])
            uf.union(parent, node)
            ancestors[uf[parent]] = parent

File: graphidx/julia/src/tree/lca.jl
import ..Utils: UnionFind, find, unite!


"""
    lowest_common_ancestors(tree, parent, pairs)

For a list of pairs `[(u1, v1), ..., (uk, vk)]` compute the
[lowest common ancestors](https://en.wikipedia.org/wiki/Lowest_common_ancestor)
`[a1, ..., ak]`, i.e. `ai` is the lowest node in the tree that is
ancestor (parent, grand-parent, grand-grand-parent, etc.) of both, `ui` and `vi`.

Runtime is `O(length(tree) + length(pairs))`.
More precisely, due to the use of the
[`UnionFind`](@ref) data structure, the inverse Ackermann function is also included;
however the impact is at most 4 for all practical instances (``n < 10^{100}``).

See CLRS (3rd edition) page 584 or
[`networkx`](https://github.com/networkx/networkx/blob/master/networkx/algorithms/lowest_common_ancestors.py).
"""
function lowest_common_ancestors(
    tree::ChildrenIndex,
    parent::Vector{Int},
    pairs::IncidenceIndex,
)::Vector{Int}
    @assert parent[root_node(tree)] == root_node(tree)

    local k = num_edges(pairs)
    local n = num_nodes(tree)
    local uf = UnionFind(n)
    local colors = fill(false, n)
    local ancestors = collect(Int, 1:n)
    lcas = fill(-1, k)

    dfs_walk(tree) do v::Int
        if v >= 0   # v just finished
            for (u, ei) in pairs[v]
                if colors[u]
                    lcas[ei] = ancestors[uf[u]]
                end
            end

            colors[v] = true
            if v != parent[v]
                unite!(uf, uf[v], uf[parent[v]])
                ancestors[uf[parent[v]]] = parent[v]
            end
        end
    end

    return lcas
end


lowest_common_ancestors(tree, parent, pairs::Vector{Tuple{Int, Int}}) =
    lowest_common_ancestors(tree, parent, IncidenceIndex(length(tree), pairs))

